Add template tag to enable derived values for string fields

rizesky opened the pull request:
Add `template: ` struct tag which renders a Go text/template using values generated for other fields. Templates are evaluated in a **second pass** after non-template fields are generated, so templates can reference provider-generated values and other template fields (**chaining supported**). Small helper functions are also provided.

**Rules**
- Only for string type fields

**Why**
Derived fields (username, email, slug, etc.) are quite common.

**Example**
```go
type UserProfile struct {
    FirstName string `faker:"first_name"`
    LastName  string `faker:"last_name"`
    Domain    string `faker:"domain_name"`

    // templates using helpers
    Username string `faker:"template:{{.FirstName | lower}}.{{.LastName | lower}}"`
    Slug     string `faker:"template:{{.FirstName | slug}}-{{.LastName | slug}}"`
    Email    string `faker:"template:{{.Username}}@{{.Domain}}"`
}

func main() {
    var u UserProfile
    if err := faker.FakeData(&u); err != nil {
        panic(err)
    }
    fmt.Printf("%+v\n", u)
}

//Result
UserProfile{
  FirstName: "John",
  LastName: "Doe",
  Domain: "example.org",
  Username: "john.doe", // lower helper applied
  Slug: "john-doe", // slug helper applied (lower + dashes)
  Email: "john.doe@example.org"// username + domain via template
}
```

rizesky committed:
feat: add template tag to enable derived values

geshtng commented on example_template_tag_test.go:
```diff
@@ -0,0 +1,34 @@
+type UserProfile struct {
+	FirstName string `faker:"first_name"`
+	LastName  string `faker:"last_name"`
+	Domain    string `faker:"domain_name"`
```
We should follow the field name or tag name?
```suggestion
	DomainName    string `faker:"domain_name"`
```
or
```suggestion
	Domain    string `faker:"domain"`
```

geshtng commented on template_tag.go:
```diff
@@ -0,0 +1,84 @@
+		return nil
+	}
+	// build context map from current struct values
+	ctx := make(map[string]interface{})
```
We can modernize by using `any`
```suggestion
	ctx := make(map[string]any)
```

geshtng commented on template_tag.go:
```diff
@@ -0,0 +1,84 @@
+
+// evaluateTemplateForField evaluates a template string tpl with context ctx (map of field name -> value)
+// and returns the resulting string. Only a small set of helper funcs are exposed.
+func evaluateTemplateForField(tpl string, ctx map[string]interface{}) (string, error) {
```
We can modernize by using `any`
```suggestion
func evaluateTemplateForField(tpl string, ctx map[string]any) (string, error) {
```

geshtng commented:
Seems chained feature still failed. 
This is the result from my local using your test code (`example_template_tag_test.go`):

<img width="801" height="81" alt="image" src="https://github.com/user-attachments/assets/9e5826f2-a4e7-417f-842d-13758ee306d4" />

rizesky committed:
Apply suggestions from code review

Co-authored-by: Gerald Sihotang <sihotangg@gmail.com>

rizesky commented:
@geshtng  Can you share how you run the example? I seem to cannot reproduce it.

geshtng commented:
> @geshtng Can you share how you run the example? I seem to cannot reproduce it.

I'm using your code `example_template_tag_test.go`. 
How to reproduce:
1. Create a new directory inside your project.
2. Copy your example code, change function `Example_templateTag` as the main function.
3. `go run main.go`
<img width="495" height="515" alt="image" src="https://github.com/user-attachments/assets/e10d5717-c349-443c-970f-db6a89dc71b3" />

geshtng commented on template_tag.go:
```diff
@@ -0,0 +1,84 @@
+		if err != nil {
+			return err
+		}
+		if v.Field(idx).Kind() != reflect.String {
+			return fmt.Errorf("template tag only supported on string fields: %s", t.Field(idx).Name)
+		}
+		v.Field(idx).SetString(res)
+	}
```
Seems you forget to update the context map:
```suggestion
		v.Field(idx).SetString(res)
		ctx[t.Field(idx).Name] = res
	}
```

rizesky replied:
you right, updated.

geshtng commented on example_template_tag_test.go:
```diff
@@ -0,0 +1,34 @@
+	"github.com/go-faker/faker/v4"
+)
+
+type Address struct {
```
Maybe we can remove this struct since its unused on this test.

rizesky replied:
removed

rizesky committed:
feat: update template-tag logic, tidy up example

geshtng commented:
Seems it's still has bug to field order.
If a template field depends on another template field that appears later in the struct, it evaluates with the dependency still empty.
This is the result using your example (just put `email` above the `username`).
<img width="479" height="398" alt="image" src="https://github.com/user-attachments/assets/afb29ace-2d2a-4117-b78d-9aa2f7c92bc3" />

geshtng commented:
You still need to check about these possible bugs:
1. Circular dependency
2. Missing field reference
3. Template tag prefix parsing

rizesky committed:
fix(template-tag): reject circular refs, missing fields, and false prefix matches

- faker.go: only treat tags of the form `template:...` as template fields,
  so bare `template` and unrelated tags like `templated` are no longer hijacked.
- template_tag.go: enable `missingkey=error` so references to undefined fields
  fail loudly instead of expanding to `<no value>`; detect self, forward, and
  circular references between template fields and return a descriptive error.
- template_tag_test.go: add regression cases for circular, self, forward, and
  missing-field references, plus a prefix-parsing test covering both the
  no-colon and false-prefix cases.

rizesky commented:
@geshtng Sorry, just have time to look at this PR again. 
I have a new commit, that I'm sure already cover are 3 possible bugs you mentioned. 
Please review

geshtng commented:
@rizesky 
Minor bug. Need to check if the field is exported or not.
Test file:
```go
func TestTemplateTagWithUnexportedField(t *testing.T) {
	type PrivateField struct {
		password  string // unexported — triggers panic in ctx-building loop
		FirstName string `faker:"first_name"`
		Slug      string `faker:"template:{{.FirstName | slug}}"`
	}

	var u PrivateField

	if err := FakeData(&u); err != nil {
		t.Fatal(err)
	}
	if u.Slug == "" {
		t.Fatal("expected non-empty Slug")
	}
}
```

Log:
<img width="745" height="439" alt="image" src="https://github.com/user-attachments/assets/300251e3-5888-457a-901e-6b96df2e65a2" />

rizesky commented:
Hi @geshtng just have time to get back to this, fixed, and I add some more comments and tests to make it safer and clearer

geshtng commented:
Wdyt? @bxcodec
